### PR TITLE
Don't inject min/max exprs in bounds of a promise_clamped of constant

### PR DESCRIPTION
**Pull request by abadams**
No effect in public Halide, but I hear it causes at least one regression
inside Google. The reason for injecting the min/max exprs in the case
where the first arg doesn't vary has since gone away now that we push
producers inside the guard-with-if if statement of a consumer.

### CODE REVIEW

File: test/correctness/CMakeLists.txt [hunk 1]
```diff
@@ -271,6 +271,7 @@ tests(GROUPS correctness travis
         skip_stages_external_array_functions.cpp
         skip_stages_memoize.cpp
         sliding_backwards.cpp
+        slicing_over_guard_with_if.cpp
```
Comment (dsharletg) [inline test/correctness/CMakeLists.txt:274]: sliding

File: src/Bounds.cpp [hunk 2]
```diff
@@ -1080,6 +1080,23 @@ class Bounds : public IRVisitor {
             Interval upper = interval;
             op->args[0].accept(this);
 
+            if (op->is_intrinsic(Call::promise_clamped) &&
+                interval.is_single_point()) {
+                // It's not safe to lift a promise_clamped
+                // intrinsic. They make a claim that holds true at
+                // that specific point in the IR. Just drop the
+                // annotation and return the bounds of the first arg.
```
Comment (dsharletg) [inline src/Bounds.cpp:1088]: ... but only when it's a single point?

File: src/Bounds.cpp [hunk 3]
```diff
@@ -1080,6 +1080,23 @@ class Bounds : public IRVisitor {
...
+                return;
+            }
+
+            if (op->is_intrinsic(Call::unsafe_promise_clamped) &&
+                interval.is_single_point(op->args[0]) &&
+                lower.is_single_point(op->args[1]) &&
+                upper.is_single_point(op->args[2])) {
+                // unsafe_promise_clamp holds globally, so we can lift the entire intrinsic.
```
Comment (dsharletg) [inline src/Bounds.cpp:1096]: Why is this true only when the operands are a single point?

File: src/Bounds.cpp [hunk 4]
```diff
@@ -1080,6 +1080,23 @@ class Bounds : public IRVisitor {
...
+                interval = Interval::single_point(op);
```
Comment (dsharletg) [inline src/Bounds.cpp:1097]: Would it be better to return `Interval::single_point(op->args[0])`?

**Comment by dsharletg** [comment]
This has fixed the performance issues I saw, and it looks like it's even making some things slightly faster than from before the regression :) But others may be getting a bit slower too.

File: src/Bounds.cpp (hunk 3 above)
Comment (steven-johnson) [reply src/Bounds.cpp:1096]: It's counterintuitive to the casual reader that promise_clamped and unsafe_promise_clamped have different global behaviors; it's probably worth adding a sentence or two clarifying why that is

**Comment by steven-johnson** [comment]
Is this ready to land (pending green bots) or does it need more pondering?

**Comment by abadams** [comment]
I'm happy with it. Up to Dillon I guess.

**Comment by dsharletg** [comment]
It's ready, and we should land it as soon as we can. Thanks for the fix!

**Comment by steven-johnson** [comment]
Will land as soon as everything goes green.

**Pull request merged**